app/code/community/Twispay/Tpay/controllers/PaymentController.php

```diff
@@ -1,362 +1,280 @@
-        
-        		}
-        	}
-        
-        	if ($result && isset($result->status) && ($result->status == 'complete-ok' || $result->status == 'in-progress')) {
-        
-        		// Set the status of this order to processing
-        		$orderId = $result->externalOrderId;
-        		$status = $result->status;
-        		Mage::log("Status: $status| Payment ID: $orderId", Zend_Log::DEBUG, $this->LOG_FILE_NAME);
-        		 
-        		try {
-        
-        			$order = Mage::getModel('sales/order')->loadByIncrementId($orderId);
-        			$order->setState(Mage_Sales_Model_Order::STATE_PROCESSING, true);
-        			$order->setStatus(Mage_Sales_Model_Order::STATE_PROCESSING);
-        			$order->addStatusToHistory($order->getStatus(), 'Order paid successfully with reference '.$result->transactionId);
-        			$order->save();
-        		} catch (LocalizedException $ex) {
-        			Mage::log($ex->getMessage(), Zend_Log::DEBUG, $this->LOG_FILE_NAME);
-        			 
-        		}
-        
-        		Mage::log("Payment has been successfully authorized. Transaction id: $result->transactionId", Zend_Log::DEBUG, $this->LOG_FILE_NAME);
-        		 
-        
-        	} else {
-        		Mage::log("Failed to complete payment for $result->transactionId", Zend_Log::DEBUG, $this->LOG_FILE_NAME);
-        		 
-        	}
-       
-        
+
+    /* Validate the decripted response. */
+    $orderValidation = Mage::helper('tpay')->twispay_tw_checkValidation($decrypted);
+
+    if(TRUE !== $orderValidation){
+      Mage::log(__FUNCTION__ . ": Failed to validate the response.", Zend_Log::ERR, $this->logFileName, /*forceLog*/TRUE);
+      $this->_redirect('checkout/onepage/failure', array ('_secure' => true));
     }
-    /**
-     * This method computes the checksum on the given data array
-     *
-     * @param string $encrypted
-     * @return array the decrypted response
-     */
-    public function decryptResponse($encrypted, $apiKey) {
-    
-    	$encrypted = (string)$encrypted;
-    	if (!strlen($encrypted)) {
-    		return null;
-    	}
-    
-    	if (strpos($encrypted, ',') !== false) {
-    		$encryptedParts = explode(',', $encrypted, 2);
-    		$iv = base64_decode($encryptedParts[0]);
-    		if (false === $iv) {
-    			throw new LocalizedException("Invalid encryption iv");
-    		}
-    		$encrypted = base64_decode($encryptedParts[1]);
-    		if (false === $encrypted) {
-    			throw new LocalizedException("Invalid encrypted data");
-    		}
-    		$decrypted = openssl_decrypt($encrypted, 'AES-256-CBC', $apiKey, OPENSSL_RAW_DATA, $iv);
-    		if (false === $decrypted) {
-    			throw new LocalizedException("Data could not be decrypted");
-    		}
-    
-    		return $decrypted;
-    	}
-    
-    	return null;
+
+    /* Extract the order. */
+    $orderId = explode('_', $decrypted['externalOrderId'])[0];
+    $order = Mage::getModel('sales/order')->loadByIncrementId($orderId);
```

Comment:
Variable `$order` is not used, is there any reason to call `loadByIncrementId`?

Reply:
I removed the line.